ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/CommandTopicBackup.java

```diff
@@ -0,0 +1,210 @@
+import io.confluent.ksql.rest.server.computation.Command;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.Pair;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Backup service that replays the KSQL command_topic to a local file. A new backup file is
+ * created whenever a new command does not match the actual backup file. Previously replayed
+ * messages read up to this new command will be written to the new file. This ensures a new
+ * complete backup of the command_topic is created.
+ */
+public class CommandTopicBackup {
+  private static final Logger LOG = LoggerFactory.getLogger(CommandTopicBackup.class);
+  private static final Ticker CURRENT_MILLIS_TICKER = new Ticker() {
+    @Override
+    public long read() {
+      return System.currentTimeMillis();
+    }
+  };
+  private static final String PREFIX = "backup_";
+
+  private final File backupLocation;
+  private final String topicName;
+  private final Ticker ticker;
+
+  private BackupReplayFile replayFile;
+  private List<Pair<CommandId, Command>> latestReplay;
+  private int latestReplayIdx;
+
+  public CommandTopicBackup(final String location, final String topicName) {
+    this(location, topicName, CURRENT_MILLIS_TICKER);
+  }
+
+  public CommandTopicBackup(final String location, final String topicName, final Ticker ticker) {
+    final File dir = new File(Objects.requireNonNull(location, "location"));
+    if (!dir.exists() || !dir.isDirectory()) {
+      throw new KsqlException(String.format(
+          "Backup location '%s' does not exist or it is not a directory.", location));
+    }
+
+    this.backupLocation = dir;
+    this.topicName = Objects.requireNonNull(topicName, "topicName");
+    this.ticker = Objects.requireNonNull(ticker, "ticker");
+  }
+
+  public void initialize() {
+    replayFile = openOrCreateReplayFile();
+
+    try {
+      latestReplay = replayFile.readRecords();
+    } catch (final IOException e) {
+      LOG.warn("Failed to read the latest backup from {}. Continue with a new file. Error = {}",
+          replayFile.getPath(), e.getMessage());
+
+      replayFile = newReplayFile();
+      latestReplay = Collections.emptyList();
+    }
+
+    latestReplayIdx = 0;
+    LOG.info("Command topic will be backup on file: {}", replayFile.getPath());
+  }
+
+  public void close() {
+    try {
+      replayFile.close();
+    } catch (final IOException e) {
+      LOG.warn("Failed closing the backup file {}. Error = {}",
+          replayFile.getPath(), e.getMessage());
+    }
+  }
+
+  @VisibleForTesting
+  BackupReplayFile getReplayFile() {
+    return replayFile;
+  }
+
+  public void writeRecord(final ConsumerRecord<CommandId, Command> record) {
```

Comment:
while the logic here is correct, it took me a while to understand what it was doing - it might make this code easier to read if you had a method `isRestoring()` and then separated the restore logic into a `checkIfRecordInLatestReplay` method

Reply:
Done.